crates/ingress-http/src/handler.rs

```diff
@@ -0,0 +1,1022 @@
+pub(crate) struct Handler<Schemas> {
+    schemas: Schemas,
+    request_tx: IngressRequestSender,
+    global_concurrency_semaphore: Arc<Semaphore>,
+}
+
+impl<Schemas> Handler<Schemas> {
+    pub(crate) fn new(
+        schemas: Schemas,
+        request_tx: IngressRequestSender,
+        global_concurrency_semaphore: Arc<Semaphore>,
+    ) -> Self {
+        Self {
+            schemas,
+            request_tx,
+            global_concurrency_semaphore,
+        }
+    }
+}
+
+impl<Schemas> Handler<Schemas>
+where
+    Schemas: ComponentMetadataResolver + Clone + Send + Sync + 'static,
+{
+    pub(crate) async fn handle<B: http_body::Body>(
+        self,
+        connect_info: ConnectInfo,
+        req: Request<B>,
+    ) -> Response<Either<Empty<Bytes>, Full<Bytes>>>
+    where
+        <B as http_body::Body>::Error: std::error::Error + Send + Sync + 'static,
+    {
+        // This chunk of code here contains CORS handling.
+        // We used to do this with tower, using https://docs.rs/tower-web/0.3.7/tower_web/middleware/cors/index.html
+        // but this is unfortunately not easy to integrate now b/c this ingress is built on hyper/http 1.0
+        // We should re-evaluate it once tower-http will provide some support for hyper/http 1.0
+
+        // Handle CORS requests
+        if req.method() == Method::OPTIONS {
+            return self.handle_preflight(req);
+        }
+
+        let result = if req.uri().path().eq_ignore_ascii_case("/restate/health") {
+            self.handle_health(req)
+        } else {
+            self.handle_inner(connect_info, req).await
+        };
+        let mut response = result.unwrap_or_else(|e| e.into_response().map(Either::Left));
+
+        response
+            .headers_mut()
+            .insert(header::ACCESS_CONTROL_ALLOW_ORIGIN, WILDCARD);
+        response
+            .headers_mut()
+            .insert(header::ACCESS_CONTROL_ALLOW_CREDENTIALS, TRUE);
+
+        response
+    }
+
+    async fn handle_inner<B: http_body::Body>(
+        self,
+        connect_info: ConnectInfo,
+        req: Request<B>,
+    ) -> Result<Response<Either<Empty<Bytes>, Full<Bytes>>>, HandlerError>
+    where
+        <B as http_body::Body>::Error: std::error::Error + Send + Sync + 'static,
+    {
+        let start_time = Instant::now();
+
+        // Acquire the semaphore permit to check if we have available quota
+        let permit = if let Ok(p) = self.global_concurrency_semaphore.try_acquire_owned() {
+            p
+        } else {
+            warn!("No available quota to process the request");
+            counter!(INGRESS_REQUESTS, "status" => REQUEST_DENIED_THROTTLE).increment(1);
+            return Err(HandlerError::TooManyRequests);
+        };
+
+        // Register request admitted
+        counter!(INGRESS_REQUESTS, "status" => REQUEST_ADMITTED).increment(1);
+
+        // Parse component_name
+        let path_parts: Vec<&str> = req.uri().path().split('/').skip(1).collect();
```

Comment:
> because at most we should have 3 path parameters?

Not so easily. It can be four if you do send on a virtual object. I also check that the user doesn't add bad paths, like `/greeter/bob/greet/thisiswrong`.

Reply:
Look at the tests here: https://github.com/restatedev/restate/pull/1236/files#diff-4dc2787df00911e7f6881c4932e2b56d1c23623901e8d753e6a70a09bf34d5adR863